core/adapters/http.go

```diff
@@ -169,54 +170,90 @@ func sendRequest(input models.RunInput, request *http.Request, config HTTPReques
 	}
 	client := &http.Client{Transport: tr}
 
-	response, err := withRetry(client, request, config)
-
+	responseBody, statusCode, err := withRetry(client, request, config)
 	if err != nil {
 		return models.NewRunOutputError(err)
 	}
 
-	defer response.Body.Close()
-
-	source := newMaxBytesReader(response.Body, config.sizeLimit)
-	bytes, err := ioutil.ReadAll(source)
-	if err != nil {
-		return models.NewRunOutputError(err)
-	}
-
-	responseBody := string(bytes)
-	if response.StatusCode >= 400 {
+	// This is either a client error caused on our end or a server error that persists even after retrying.
+	// Either way, there is no way for us to complete the run with a result.
+	if statusCode >= 400 {
 		return models.NewRunOutputError(errors.New(responseBody))
 	}
 
 	return models.NewRunOutputCompleteWithResult(responseBody)
 }
 
+// withRetry executes the http request in a retry. Timeout is controlled with a context
+// Retry occurs if the request timeout, or there is any kind of connection or transport-layer error
+// Retry also occurs on remote server 5xx errors
 func withRetry(
 	client *http.Client,
 	originalRequest *http.Request,
 	config HTTPRequestConfig,
-) (*http.Response, error) {
-	var response *http.Response
-	err := retry.Do(
+) (responseBody string, statusCode int, err error) {
+	err = retry.Do(
 		func() error {
 			ctx, cancel := context.WithTimeout(context.Background(), config.timeout)
 			defer cancel()
 			requestWithTimeout := originalRequest.Clone(ctx)
 
+			start := time.Now()
+
 			r, err := client.Do(requestWithTimeout)
 			if err != nil {
 				return err
 			}
-			response = r
+			defer r.Body.Close()
+			statusCode = r.StatusCode
+			elapsed := time.Since(start)
+			logger.Debugw(fmt.Sprintf("http adapter got %v in %s", statusCode, elapsed), "statusCode", statusCode, "timeElapsedSeconds", elapsed)
+
+			source := newMaxBytesReader(r.Body, config.sizeLimit)
+			bytes, err := ioutil.ReadAll(source)
+			if err != nil {
+				logger.Errorf("http adapter error reading body: %v", err.Error())
+				return err
+			}
+			elapsed = time.Since(start)
+			logger.Debugw(fmt.Sprintf("http adapter finished after %s", elapsed), "statusCode", statusCode, "timeElapsedSeconds", elapsed)
+
+			responseBody = string(bytes)
```

Comment:
This is a little weird, directly casting []bytes to strings, we may want to leave it as `[]byte` as long as possible or at least until `json` decoding as this seems to just directly chop out invalid utf-8 and leave the rest intact, which is a bit unexpected.

Reply:
Yeah can do, it was a tiny optimisation since we only ever use it as a string and I needed to stringify it for the logging.

Reply:
Changed it :+1: